### PR TITLE
Deprecate `absubmit` and `acousticbrainz` plugins

### LINKED ISSUE #4627: acousticbrainz and absubmit plugins should be deprecated
Acousticbrainz has [shut down](https://blog.metabrainz.org/2022/02/16/acousticbrainz-making-a-hard-decision-to-end-the-project/) and will go completely offline in early 2023 according to that blog post. The `acousticbrainz` and `absubmit` plugins should be deprecated.

Attempting to use it now just results in a lot of "acousticbrainz did not provide info about..." messages.

### PR DESCRIPTION
**Pull request by trolley**
## Description

Fixes #4627.

AcousticBrainz is shutting down as of early 2023. Deprecate the `absubmit` and `acousticbrainz` plugins, and modify them to only function if an AcousticBrainz server has been configured.

## To Do

- [x] Documentation. (If you've add a new command-line flag, for example, find the appropriate page under `docs/` to describe it.)
- [x] Changelog. (Add an entry to `docs/changelog.rst` near the top of the document.)
- [x] Tests. (Encouraged but not strictly required.)

### CODE REVIEW

**Comment by JOJ0** [comment]
@trolley I don't see a problem with finally merging this one. Would you be so kind and resolve conflicts and mark remaining checkmarks.

Could you post how this deprecation now would present itself to a beets user?

File: beetsplug/acousticbrainz.py
Comment (JOJ0) [inline beetsplug/acousticbrainz.py:42]: Maybe we could be a little more elaborate here. Something similar to what you've put in the changelog sounds a little nicer:

This plugin is deprecated and no longer has any function since the AcousicBrainz project has been shut down.

**Comment by trolley** [comment]
Given a config file with `absubmit` and/or `acousticbrainz` plugins in it, like:
```
plugins: absubmit acousticbrainz

absubmit:
  auto: yes
```
These messages are printed to the user when beets is invoked:
```
(beets-dev)  π ~ ❯ beet ls
acousticbrainz: This plugin is deprecated and no longer functions.
absubmit: This plugin is deprecated and no longer functions.
```

If either plugin is manually invoked the same messages appear, along with the longer message:
```
(beets-dev)  π ~ ❯ beet absubmit
acousticbrainz: This plugin is deprecated and no longer functions.
absubmit: This plugin is deprecated and no longer functions.
This plugin is deprecated and no longer has any function since AcousticBrainz no longer accepts new submissions.
(beets-dev)  π ~ ❯ beet acousticbrainz
acousticbrainz: This plugin is deprecated and no longer functions.
absubmit: This plugin is deprecated and no longer functions.
This plugin is deprecated and no longer has any function since the AcousicBrainz project has been shut down.
```

**Comment by Serene-Arc** [comment]
`asubmit` is likely pointless moving forwards but can't there be a local version of acousticbrains? Such as in a docker instance? It might be a little premature to say that the entire plugin is deprecated to the point of uselessness.

**Comment by trolley** [comment]
If the maintainers decide they'd prefer that I will adjust the PR.

Personally, I don't see much point considering the maintainers of AcousticBrainz themselves say:

> Unfortunately, a number of things happened with the data that we collected which made us decide that the quality of the data isn’t as useful as we had hoped.

**Comment by JOJ0** [comment]
> `asubmit` is likely pointless moving forwards but can't there be a local version of acousticbrains? Such as in a docker instance? It might be a little premature to say that the entire plugin is deprecated to the point of uselessness.

There is a local version of this plugin already that does everything absubmit die and more. It is a third party plugin but I'm co-maintaining it: https://github.com/adamjakab/BeetsPluginXtractor

**Comment by JOJ0** [comment]
@Serene-Arc @trolley and I'd like to include @sampsyo into this discussion: Now after having thought about this a little more I'd like to go one step further and even add to the deprecation message that it should be considered to move from acousticbrainz plugin to BeetsPluginXtractor. 

If adding it to the deprecation message is not decided to be the best way, we could also add it to the acousticbrainz plugins docs page: https://beets.readthedocs.io/en/stable/plugins/acousticbrainz.html

What do you think about this idea? I would hope to get some movement to the third party plugin.

**Comment by trolley** [comment]
Output with latest change:
```
(beets-dev)  π ~ ❯ beet absubmit
acousticbrainz: This plugin is deprecated and no longer functions.
absubmit: This plugin is deprecated and no longer functions.
The AcousticBrainz project has been shut down and thus no longer accepts new submissions.
(beets-dev)  π ~ ❯ beet acousticbrainz
acousticbrainz: This plugin is deprecated and no longer functions.
absubmit: This plugin is deprecated and no longer functions.
The AcousicBrainz project has been shut down.
```

**Comment by JOJ0** [comment]
@sampsyo any last wording advices for this deprecation message? Sorry for bothering, I just feel a little afraid while having to do a final decision for kicking out a part of beets that's  been in there for years and did good work. Even if it's of no use any more. Help! Thanks so much in advance! :-)

**Comment by sampsyo** [comment]
The deprecation messages look great; thank you so much to everybody for iterating on them! Short and simple and clear.

I'll defer to others about whether it's the right thing to delete the code (vs. just printing the deprecation message). As @Serene-Arc implies, if there's any chance people have their own local thing running, then maybe that's a reason to keep the code. But as @JOJ0 mentions also, `BeetsPluginXtractor` is clearly the spiritual successor.

**Comment by JOJ0** [comment]
@Serene-Arc are you fine with kicking out the code or should we keep it? Your idea of "dockerizing" AB and keeping this plug as-is (in general) is a different approach to the xtractor plug. On the other hand, git never forgets so it's not lost forever. Your call, I'm fine either way! :-) Thanks!

**Comment by Serene-Arc** [comment]
I think that, if the code isn't completely useless, which it isn't because there are docker services, then there's no real cost to keeping it in the codebase. It might disrupt a few workflows to remove it and if there's no real reason to remove it, then I think it should be kept.

The deprecation message will make sure that people know to switch if they're not explicitly running their own service, and moving forwards we can discuss whether to decline any further development on the plugins.

**Comment by zaptrem** [comment]
Fuzzy but sort-of correct data like this is extremely useful to ML researchers working on music generation.

**Comment by JOJ0** [comment]
> I think that, if the code isn't completely useless, which it isn't because there are docker services, then there's no real cost to keeping it in the codebase. It might disrupt a few workflows to remove it and if there's no real reason to remove it, then I think it should be kept.

Alright then let's keep it :-) I would have believed in git, saving everything for us until the end of time but certainly if this plugin stays here, it's not hidedn in old commits and the chance someone picks up the idea and creates a dockerized versoin of AcousticBrainz might be more likely.

@trolley sorry for the hassle but would you be so kind and bring back the deleted stuff. I hope it's not too much of a hassle now :-)

> The deprecation message will make sure that people know to switch if they're not explicitly running their own service, and moving forwards we can discuss whether to decline any further development on the plugins.

Should deprecation messages be altered now or are they fine like that? Not sure....

**Comment by JOJ0** [comment]
> Fuzzy but sort-of correct data like this is extremely useful to ML researchers working on music generation.

That's out of question! Yes! And just to clarify: The deprecation of this defunct plugin is not done because we don't think the outcome isn't useful because someone (MB) said that!.

I still think though, that the alternative that currently _is_ stable and should be supported is https://github.com/adamjakab/BeetsPluginXtractor. You might have missed that in the long threads above but it uses _exactly_ the algorithms / the software-stack AcousticBrainz was using, only that it allows to even use _newer_ versions of that software (https://github.com/mtg/essentia) that might be of even more use because they might have fixed some bugs AB was facing in the meantime.v I think that's awesome and my personal experience is now even better since I started to use xtractor instead of acousticbrainz plugin. The main problem I had with acousticbrainz was that every now and then some files could never be analysed and it was kind of a blackbox what the problem was. Was it my file, was it a server error over at acousticbrainz.org, I never new. Now everything under control because it runs locally with newer versions and debug logging if I want that.

**Comment by trolley** [comment]
I restored the acousticbrainz plugin but added a new config option `hostname` that's required.

I tested it with a couple scenarios:

- without the `hostname` config set:

```
(beets-dev)  π ~ ❯ beet acousticbrainz
error: AcousticBrainz server hostname not specified
```

- with the `hostname` config set, but to a hostname that doesn't work:

```
(beets-dev)  π ~ ❯ beet acousticbrainz
acousticbrainz: getting data for: Thrice - Horizons / East - The Color of the Sky
acousticbrainz: request error: HTTPConnectionPool(host='fakeserver', port=80): Max retries exceeded with url: /a24cd93d-7a48-4971-90de-38f628a7a8ff/low-level (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f1cb1b38eb0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution'))
```

I'll try to set up my own docker instance of AcousticBrainz to try it with that when I can.

File: beetsplug/acousticbrainz.py
Comment (trolley) [inline beetsplug/acousticbrainz.py:149]: I'm considering a couple things about this:
- I think this option should be renamed to `base_url` since it needs to include the scheme and possibly port.
- Should more validation happen here, like verifying the option _does_ include a scheme? Only `http` or `https` would be supported.

**Comment by trolley** [comment]
I had no luck building their docker image. Could someone else give it a try?

File: docs/plugins/absubmit.rst
Comment (JOJ0) [inline docs/plugins/absubmit.rst:8]: Since we decided the plugin should stay and allow future tinkering with dockerised self-hosting, I suggest we bring back all the docs above this line (including the deprecation message).

Also I would now propse that since we have BeetsPluginXtractor in the community-plugins list already (https://beets.readthedocs.io/en/stable/plugins/index.html#other-plugins) I'd like to include a sentence next to the deprecation message that says:

As an alternative to the `absubmit` and `acouticbrainz` plugins ´[beets-xtractor](https://github.com/adamjakab/BeetsPluginXtractor)` can be used

File: docs/plugins/index.rst
Comment (JOJ0) [inline docs/plugins/index.rst:160]: Also this should be brought back I suppose. Since a dockerized version would require to have exactly the same system of using `absubmit` to analyze and push to server/container and `acousticbrainz` then to retrieve it back from there.

**Review by JOJ0 — CHANGES_REQUESTED** [review]
The goals of this PR from my point of view would be:

- Print deprecation messages in both plugins. DONE
- Allow future tinkering / containerization. Code should be kept in-tact.
- Proposal: Mention beets-xtractor plugin as a currently working replacement in the docs of both plugins.

Non-goals IMHO:
- Provide a functioning container version.
- Alter the plugin further to improve those future ideas.

We certainly can leave the hostname feature you already implemented @trolley , it is a good idea and a nice start for future tinkerers! :-)

@serene-arc, what do you think about setting these goals to finalise this PR?

**Comment by trolley** [comment]
Alright, I took another crack at it. Now both plugins remain functional if you provide a configuration option to set the server. Both still print a deprecation warning if included in a config file, and a more detailed deprecation message pointing to the new config option of explicitly invoked.

File: beetsplug/absubmit.py
Comment (trolley) [inline beetsplug/absubmit.py:85]: The download links on the AcousticBrainz site are dead now, so I updated this to the official Essentia project homepage where it's still available.

Comment (trolley) [reply beetsplug/absubmit.py:85]: Actually it's only the 64-bit Linux version that's dead, but it's probably still a good idea to reference the official project site instead of the mirrored downloads on the AcousticBrainz site which may not be around for long.

**Review by JOJ0 — APPROVED** [review]
Thanks for the changes in wording and link changes. That is much appreciated and very thoughtful! Great! And many thanks for the patience since it took some time to really know where and how this deprecation should be going. I think it was all worth it and I am finally....merging :bomb: :

**Comment by duckfromdiscord** [comment]
I have a quick question about this PR, feel free to hide my message if it seems unimportant.

I wanted to thank you for including the custom server option. My question is, are there any custom servers out there we can point the plugin to?

**Comment by JOJ0** [comment]
Not that I'm aware of. FWIW, have you seen the suggested alternative for getting similar information into your library noted in the second paragraph here: https://beets.readthedocs.io/en/latest/plugins/acousticbrainz.html

**Comment by duckfromdiscord** [comment]
`beets-xtractor`? I've seen it also, I was going to ask there but I noticed they didn't have a base URL option. I wanted to see if there was an online dataset somewhere that would include analyses of other music. Thank you for the help.

**Comment by JOJ0** [comment]
xtractor does the analysis on your machine ☝️😁

**Pull request merged**
